compiler/rustc_passes/src/liveness.rs

```diff
@@ -329,6 +329,48 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
             }
         }
 
+        // Allow todo! macro
+        /*
+            Skips checking for unused variables when the trailing expression
+            of the body is a panic with a message that contains "not yet implemented".
+
+            # Example
+
+            fn foo(x: i32) {
+                // arbitrary code
+                todo!()
+            }
+        */
+        if let ExprKind::Block(block, _) = &body.value.kind {
+            if let Some(expr) = block.expr {
+                if let ExprKind::Call(call, [arg]) = expr.kind {
+                    if let ExprKind::Path(QPath::Resolved(_, path)) = call.kind {
+                        if let Res::Def(DefKind::Fn, path_def_id) = &path.res {
+                            let panic_fn = self.tcx.lang_items().panic_fn();
+                            // Note: there is no function for panic_fmt, so we have to extract it from the debug output :(
+                            // builder doesn't like this being called without panic `self.tcx.def_path_str(*path_def_id);`
+                            let path_str = format!("{:?}", path_def_id);
+                            if Some(*path_def_id) == panic_fn
+                                || ((path_str.contains("std[") || path_str.contains("core["))
```

Comment:
By the way, why does this use `[`?

Reply:
The path_str is like `core[76aa]::panicking::panic`. I was concerned about matching on some arbitrary crate with a matching path, though it's a bit unlikely. I actually think it should be `~ core[" .. "]::panicking_panic" ..` just to ensure that it only triggers for std / core.